File: rust/threadhop-core/src/digest.rs
//! Per-session digest — the data shape behind the right-hand digest
//! panel.
//!
//! **Pre-pop scaffolding.** Worker H fills in the real
//! [`compute_session_digest`] builder in the deferrals-cleanup wave by
//! porting `threadhop_core/session/digest.py::extract_digest` to Rust.
//! Until then, the stub returns a digest with only `session_id` and a
//! short truncated `slug` populated so callers can render an "empty
//! digest" panel without crashing.
//!
//! The struct field set mirrors the Python `SessionDigest` dataclass —
//! the computed-property names from Python (`title`, `cache_hit_ratio`,
//! `context_fill_ratio`, `files_touched_count`,
//! `total_input_tokens_billed`) are flattened into regular fields here
//! because Rust doesn't have lazy properties on data structs. Worker H
//! computes those values during the JSONL pass and stores them
//! directly.

use rusqlite::Connection;

/// One band of the recap timeline (Started / Earlier / Recently / Last
/// asked). `timestamp` is the source ISO timestamp where available; the
/// renderer uses it to compute a relative-age suffix.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct RecapEntry {
    pub label: String,
    pub timestamp: Option<String>,
    pub text: String,
}

/// A single session, summarised for the right-column digest panel.
///
/// Worker H populates this from a single pass over the JSONL plus a
/// small SQLite lookup for resume-command metadata. Pre-pop only
/// guarantees `session_id` and a slug derived from it.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct SessionDigest {
    /// Stable session UUID.
    pub session_id: String,

    /// Best display title — custom title > ai title > slug > truncated
    /// session id. Empty until Worker H runs.
    pub title: String,

    /// Short identifier — typically the first 8 chars of `session_id`,
    /// or a user-supplied slug.
    pub slug: String,

    /// Git branch the session was opened on, if recorded.
    pub branch: Option<String>,

    /// Wall-clock duration (seconds) from first user prompt to last
    /// event observed.
    pub duration_seconds: Option<u64>,

    /// Recap bands in display order — empty when nothing is available
    /// yet (fresh sessions with no completed turn).
    pub recap: Vec<RecapEntry>,

    /// PR number captured from a `pr-link` line, if any.
    pub pr_number: Option<u64>,

    /// Repository the PR lives in (e.g. `owner/repo`).
    pub pr_repository: Option<String>,

    /// Distinct file paths the assistant edited / wrote during the
    /// session. Sorted, deduped.
    pub files_touched: Vec<String>,

    /// `files_touched.len() as u64`, precomputed so the renderer
    /// doesn't have to call `.len()` on a Vec it doesn't own.
    pub files_touched_count: u64,

    /// Context window in tokens for the most recent assistant model.
    /// `None` when no assistant turn has reported usage yet.
    pub context_window: Option<u64>,

    /// Input-tokens count for the latest assistant turn — drives the
    /// "context fill" gauge.
    pub latest_turn_input_tokens: u64,

    /// Cumulative TOTAL input tokens across the session (uncached +
    /// cache_read + cache_creation). Matches the Python
    /// `total_input_tokens_billed` computed property.
    pub total_input_tokens_billed: u64,

    /// Cumulative output tokens.
    pub total_output_tokens: u64,

    /// `latest_turn_input_tokens / context_window`, clamped to 1.0.
    /// `None` until at least one assistant turn has run.
    pub context_fill_ratio: Option<f32>,

    /// `cache_read / (cache_read + cache_creation)`. `None` when the
    /// denominator is zero.
    pub cache_hit_ratio: Option<f32>,

    /// Models actually used in the session, in first-seen order.
    pub models_used: Vec<String>,

    /// Permission mode captured from a `permission-mode` line.
    pub permission_mode: Option<String>,

    /// Client version captured from any line carrying a `version`
    /// field.
    pub client_version: Option<String>,
}

/// Build a session digest. **Pre-pop stub.** Worker H replaces the body
/// with a real JSONL pass; until then this returns a digest with just
/// `session_id` and a slug derived from the first 8 chars of the id.
///
/// The signature is fixed: workers and tests can wire calls without
/// risking a churn when the real implementation lands.
pub fn compute_session_digest(session_id: &str, _conn: &Connection) -> SessionDigest {
    SessionDigest {
        session_id: session_id.to_string(),
        slug: session_id.chars().take(8).collect(),
        ..Default::default()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn stub_populates_session_id_and_slug() {
        let conn = Connection::open_in_memory().unwrap();
        let digest = compute_session_digest(
            "abcdef1234567890-rest-of-uuid",
            &conn,
        );
        assert_eq!(digest.session_id, "abcdef1234567890-rest-of-uuid");
        // Slug is the first 8 chars of the session id — matches the
        // Python `title` fallback semantics when no custom title /
        // ai title / slug field is set.
        assert_eq!(digest.slug, "abcdef12");
    }

    #[test]
    fn stub_leaves_other_fields_default() {
        let conn = Connection::open_in_memory().unwrap();
        let digest = compute_session_digest("session-id", &conn);
        // Pre-pop guarantee: every other field is Default. This protects
        // Worker H from accidentally relying on side-effects from the
        // stub when wiring real population.
        assert!(digest.title.is_empty());
        assert_eq!(digest.branch, None);
        assert_eq!(digest.duration_seconds, None);
        assert!(digest.recap.is_empty());
        assert_eq!(digest.pr_number, None);
        assert!(digest.files_touched.is_empty());
        assert_eq!(digest.files_touched_count, 0);
        assert_eq!(digest.context_window, None);
        assert_eq!(digest.latest_turn_input_tokens, 0);
        assert_eq!(digest.total_input_tokens_billed, 0);
        assert_eq!(digest.total_output_tokens, 0);
        assert_eq!(digest.context_fill_ratio, None);
        assert_eq!(digest.cache_hit_ratio, None);
        assert!(digest.models_used.is_empty());
        assert_eq!(digest.permission_mode, None);
        assert_eq!(digest.client_version, None);
    }

    #[test]
    fn slug_handles_short_session_id() {
        let conn = Connection::open_in_memory().unwrap();
        let digest = compute_session_digest("abc", &conn);
        // Slug never panics on a short id — just takes whatever's
        // there. Worker H may eventually override this from a real
        // `slug` line in the JSONL.
        assert_eq!(digest.slug, "abc");
    }

    #[test]
    fn recap_entry_default_is_empty() {
        let entry = RecapEntry::default();
        assert!(entry.label.is_empty());
        assert!(entry.text.is_empty());
        assert_eq!(entry.timestamp, None);
    }
}

File: threadhop_core/session/digest.py
"""Per-session digest extracted from a Claude Code JSONL transcript.

The digest is the data shape behind the right-hand sidebar in the TUI
(``SessionDigestBar``). It is a single read-only snapshot of everything
the bar wants to show, computed in one pass over the JSONL.

Why a separate module rather than extending the indexer:

- The indexer's job is the FTS view of conversation lines. It deliberately
  drops every non-message line type (``_NON_MESSAGE_TYPES`` in ``models.py``).
  The digest reads exactly those dropped lines (``ai-title``, ``custom-title``,
  ``pr-link``, ``last-prompt``, ``permission-mode``, ``system`` subtypes),
  plus a few ambient fields on user/assistant lines (``slug``, ``gitBranch``,
  ``version``, ``message.model``, ``message.usage``).
- Keeping the extractor outside ``tui/`` lets the HTTP sidecar reuse it
  later without pulling Textual into ``server/app.py``.

The four-band recap timeline (Started → Earlier → Recently → Last asked)
is built here too, because it depends on the same single-pass read.
"""

from __future__ import annotations

import json
import logging
import re
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path

log = logging.getLogger("threadhop.session.digest")


# Which tool_use names imply a file was touched. Edit/Write/MultiEdit cover
# the common cases; NotebookEdit is rare but cheap to include.
_FILE_TOUCH_TOOLS = frozenset({"Edit", "Write", "MultiEdit", "NotebookEdit"})

# Default context window when we can't infer one from the model id.
# Every Claude 4.x model is ≥200k; the [1m] suffix on opus-4-7 lifts it
# to 1M. We don't try to enumerate every model — just spot the one
# distinguishing flag and let the rest fall through to the safe default.
_DEFAULT_CONTEXT_WINDOW = 200_000


def _model_context_window(model: str | None) -> int:
    """Return the context window size in tokens for a given model id."""
    if not model:
        return _DEFAULT_CONTEXT_WINDOW
    if "[1m]" in model.lower() or model.lower().endswith("-1m"):
        return 1_000_000
    return _DEFAULT_CONTEXT_WINDOW


# Regexes for cleaning the first user prompt before it becomes the
# "Started" recap band. The patterns target the agent-injected
# envelopes that show up in real transcripts (system reminders,
# task-notification XML, "[From …]" previews) — they stay anchored
# at start-of-string so they only fire on prefix noise, never on
# legitimate body text.
_NOISE_PREFIX_PATTERNS = [
    re.compile(r"^\s*<task-notification>.*?</task-notification>\s*", re.DOTALL),
    re.compile(r"^\s*<system-reminder>.*?</system-reminder>\s*", re.DOTALL),
    re.compile(r"^\s*\[From\s+\".*?\"\s*[—-]\s*[^\]]*\]\s*"),
]

# Maximum length of the cleaned Started band. Long enough for a real
# question, short enough that the body never overflows two visual rows
# at the bar's 36-cell width.
_STARTED_MAX_CHARS = 140


def _clean_first_prompt(text: str) -> str:
    """Strip agent-injected envelope noise and take a single-sentence headline.

    Real transcripts often start with ``<task-notification>`` XML, a
    ``[From "..." — path]`` preview, or other agent scaffolding. Those
    prefixes don't read as a useful "what did the user originally ask"
    headline; they're just framing. We strip the known patterns, then
    take the first non-empty line, then truncate at the first sentence
    boundary or 140 chars — whichever comes first.
    """
    cleaned = text.strip()
    for pattern in _NOISE_PREFIX_PATTERNS:
        cleaned = pattern.sub("", cleaned, count=1).strip()

    # First non-empty line.
    for line in cleaned.splitlines():
        if line.strip():
            cleaned = line.strip()
            break

    # First sentence — split on the earliest of '.', '?', '!' if it ends
    # with whitespace (so URLs and version numbers don't trigger).
    match = re.search(r"[.!?](?:\s|$)", cleaned)
    if match and match.start() < _STARTED_MAX_CHARS:
        cleaned = cleaned[: match.end()].rstrip()
    if len(cleaned) > _STARTED_MAX_CHARS:
        # Cut at the last word boundary before the cap so we don't
        # truncate mid-word.
        snippet = cleaned[:_STARTED_MAX_CHARS]
        space = snippet.rfind(" ")
        if space > _STARTED_MAX_CHARS - 30:
            snippet = snippet[:space]
        cleaned = snippet.rstrip(" ,;:.") + "…"
    return cleaned


@dataclass
class RecapEntry:
    """One band of the recap timeline.

    ``label`` is the band name shown in the bar (``"Started"``,
    ``"Earlier"``, ``"Recently"``, ``"Last asked"``). ``timestamp`` is the
    line's ISO timestamp where available — used for the small relative
    age suffix. ``text`` is the recap body, untruncated; the widget
    decides how many lines to show.
    """

    label: str
    timestamp: str | None
    text: str


@dataclass
class SessionDigest:
    """One session, summarized for the sidebar in a single pass."""

    session_id: str

    # Identity / context line
    custom_title: str | None = None
    ai_title: str | None = None
    slug: str | None = None
    branch: str | None = None
    cwd: str | None = None

    # Wall-clock duration in seconds, computed from first to last
    # message timestamps observed in the JSONL.
    duration_seconds: int | None = None

    # Recap bands, in display order. Empty when nothing is available.
    recap: list[RecapEntry] = field(default_factory=list)

    # Outputs
    pr_number: int | None = None
    pr_url: str | None = None
    pr_repository: str | None = None
    files_touched: set[str] = field(default_factory=set)

    # Token usage (sums across the session). Cache hit ratio is computed
    # by the consumer from cache_read / (cache_read + cache_creation).
    total_input_tokens: int = 0
    total_output_tokens: int = 0
    total_cache_read_tokens: int = 0
    total_cache_creation_tokens: int = 0

    # The size of the prompt at the *latest* assistant turn — the
    # number that drives the "context fill" gauge in the bar.
    # Computed as latest_usage.input_tokens + cache_read + cache_creation
    # because all three are part of the prompt the model saw on that
    # turn. Cumulative session totals live in the ``total_*`` fields
    # above and are independent of context-window pressure.
    latest_turn_input_tokens: int = 0

    # Context window in tokens for the *latest* assistant model.
    # Defaults to 200k; bumped to 1M when the model id carries the
    # ``[1m]`` suffix (opus-4-7 long-context). The bar uses this as
    # the denominator of the context-fill ratio.
    context_window: int = _DEFAULT_CONTEXT_WINDOW

    # Models actually used in this session, in first-seen order. Useful
    # when a session has both Opus turns (chat) and Haiku turns
    # (skill subcalls).
    models_used: list[str] = field(default_factory=list)

    # Footer
    permission_mode: str | None = None
    client_version: str | None = None

    @property
    def files_touched_count(self) -> int:
        return len(self.files_touched)

    @property
    def cache_hit_ratio(self) -> float | None:
        denom = self.total_cache_read_tokens + self.total_cache_creation_tokens
        if denom == 0:
            return None
        return self.total_cache_read_tokens / denom

    @property
    def total_tokens(self) -> int:
        return (
            self.total_input_tokens
            + self.total_output_tokens
            + self.total_cache_read_tokens
            + self.total_cache_creation_tokens
        )

    @property
    def total_input_tokens_billed(self) -> int:
        """Cumulative TOTAL input billed across the session.

        ``total_input_tokens`` alone is just the new (uncached) input;
        the cache-read and cache-creation tokens are also part of every
        turn's prompt. The headline "input" number on the bar should
        be all three combined so the user sees the real input weight,
        not just the uncached slice.
        """
        return (
            self.total_input_tokens
            + self.total_cache_read_tokens
            + self.total_cache_creation_tokens
        )

    @property
    def context_fill_ratio(self) -> float | None:
        """Latest turn's prompt size as a fraction of the context window.

        ``None`` when no assistant turn has reported usage yet. Capped
        at 1.0 — sessions that touch compaction can briefly exceed the
        window before the compactor lands, and a "103%" reading is
        more confusing than informative on a passive bar.
        """
        if self.latest_turn_input_tokens <= 0 or self.context_window <= 0:
            return None
        ratio = self.latest_turn_input_tokens / self.context_window
        return min(ratio, 1.0)

    @property
    def title(self) -> str:
        """Best display title: custom > ai > slug > truncated session id."""
        if self.custom_title:
            return self.custom_title
        if self.ai_title:
            return self.ai_title
        if self.slug:
            return self.slug
        return self.session_id[:8]


def _parse_iso(ts: str | None) -> datetime | None:
    if not ts:
        return None
    try:
        return datetime.fromisoformat(str(ts).replace("Z", "+00:00"))
    except Exception:
        return None


def _truncate_lines(text: str, max_lines: int) -> str:
    lines = text.splitlines()
    if len(lines) <= max_lines:
        return text
    return "\n".join(lines[:max_lines]).rstrip() + "\n…"


def extract_digest(jsonl_path: Path, *, session_id: str | None = None) -> SessionDigest:
    """Single-pass extractor. Returns a partial digest if the file errors mid-read.

    ``session_id`` defaults to the file stem when not provided. Lines that
    fail to decode are skipped silently (matches indexer behavior); lines
    with shapes we don't recognize are simply ignored.
    """
    sid = session_id or jsonl_path.stem
    digest = SessionDigest(session_id=sid)

    first_user_text: str | None = None
    first_user_ts: str | None = None
    last_event_ts: str | None = None

    away_summary_text: str | None = None
    away_summary_ts: str | None = None
    compact_summary_text: str | None = None
    compact_summary_ts: str | None = None

    # Dedupe assistant usage by message.id (chunks share an id; usage
    # lives on the final chunk and last-write-wins gives us the totals).
    usage_by_msg_id: dict[str, dict] = {}
    seen_models: list[str] = []
    seen_models_set: set[str] = set()

    # Track the latest assistant turn to compute context fill against
    # that model's window. ``latest_usage`` mirrors the final usage
    # block we saw; ``latest_model`` is whichever model id rode that
    # final assistant message.
    latest_usage: dict | None = None
    latest_model: str | None = None

    try:
        with jsonl_path.open() as fh:
            for line in fh:
                line = line.strip()
                if not line:
                    continue
                try:
                    obj = json.loads(line)
                except json.JSONDecodeError:
                    continue
                if not isinstance(obj, dict):
                    continue

                t = obj.get("type")

                # Ambient fields present on most line types.
                if obj.get("slug") and not digest.slug:
                    digest.slug = str(obj["slug"])
                if obj.get("gitBranch"):
                    digest.branch = str(obj["gitBranch"])
                if obj.get("version"):
                    digest.client_version = str(obj["version"])
                if obj.get("cwd") and not digest.cwd:
                    digest.cwd = str(obj["cwd"])
                ts = obj.get("timestamp")
                if isinstance(ts, str):
                    last_event_ts = ts

                if t == "custom-title":
                    val = obj.get("customTitle")
                    if val:
                        digest.custom_title = str(val)
                elif t == "ai-title":
                    val = obj.get("aiTitle")
                    if val:
                        digest.ai_title = str(val)
                elif t == "pr-link":
                    pr_num = obj.get("prNumber")
                    if pr_num is not None:
                        try:
                            digest.pr_number = int(pr_num)
                        except (TypeError, ValueError):
                            pass
                    if obj.get("prUrl"):
                        digest.pr_url = str(obj["prUrl"])
                    if obj.get("prRepository"):
                        digest.pr_repository = str(obj["prRepository"])
                elif t == "permission-mode":
                    val = obj.get("permissionMode")
                    if val:
                        digest.permission_mode = str(val)
                elif t == "system":
                    subtype = obj.get("subtype")
                    content = obj.get("content")
                    if subtype == "away_summary" and isinstance(content, str) and content:
                        away_summary_text = content
                        away_summary_ts = ts if isinstance(ts, str) else away_summary_ts
                    elif subtype == "compact_boundary":
                        cm = obj.get("compactMetadata") or {}
                        # The compact_boundary line itself only carries
                        # "Conversation compacted" — the actual digest is
                        # written as the next user line with
                        # isCompactSummary=true. We capture the timestamp
                        # here; the text is filled in below.
                        if isinstance(ts, str):
                            compact_summary_ts = ts
                        # Fallback: if a future schema ever inlines the
                        # summary on the boundary line, take it.
                        if isinstance(content, str) and len(content) > 30:
                            compact_summary_text = content
                elif t == "user":
                    msg = obj.get("message") or {}
                    content = msg.get("content")
                    is_compact = bool(obj.get("isCompactSummary"))
                    is_tool_result = bool(obj.get("toolUseResult"))
                    if is_compact and isinstance(content, str):
                        compact_summary_text = content
                        if isinstance(ts, str):
                            compact_summary_ts = ts
                    elif not is_tool_result and isinstance(content, str) and content:
                        if first_user_text is None:
                            first_user_text = content
                            if isinstance(ts, str):
                                first_user_ts = ts
                elif t == "assistant":
                    msg = obj.get("message") or {}
                    model = msg.get("model")
                    if isinstance(model, str) and model:
                        if model not in seen_models_set:
                            seen_models.append(model)
                            seen_models_set.add(model)
                        latest_model = model
                    msg_id = msg.get("id")
                    usage = msg.get("usage")
                    if isinstance(msg_id, str) and isinstance(usage, dict):
                        usage_by_msg_id[msg_id] = usage
                        latest_usage = usage
                    content = msg.get("content")
                    if isinstance(content, list):
                        for block in content:
                            if not isinstance(block, dict):
                                continue
                            if block.get("type") != "tool_use":
                                continue
                            if block.get("name") not in _FILE_TOUCH_TOOLS:
                                continue
                            inp = block.get("input") or {}
                            if not isinstance(inp, dict):
                                continue
                            fp = inp.get("file_path") or inp.get("notebook_path")
                            if isinstance(fp, str) and fp:
                                digest.files_touched.add(fp)
    except FileNotFoundError:
        log.debug("digest extractor: %s not found", jsonl_path)
        return digest
    except Exception as e:
        log.warning("digest extractor: partial read of %s — %s", jsonl_path, e)

    # Sum token usage across deduped messages.
    for u in usage_by_msg_id.values():
        digest.total_input_tokens += int(u.get("input_tokens") or 0)
        digest.total_output_tokens += int(u.get("output_tokens") or 0)
        digest.total_cache_read_tokens += int(u.get("cache_read_input_tokens") or 0)
        digest.total_cache_creation_tokens += int(
            u.get("cache_creation_input_tokens") or 0
        )

    digest.models_used = seen_models

    # Latest assistant turn → context fill. Sum input + cache_read +
    # cache_creation because all three count against the prompt size
    # the model saw. The model id picks the context window denominator.
    if isinstance(latest_usage, dict):
        digest.latest_turn_input_tokens = (
            int(latest_usage.get("input_tokens") or 0)
            + int(latest_usage.get("cache_read_input_tokens") or 0)
            + int(latest_usage.get("cache_creation_input_tokens") or 0)
        )
    digest.context_window = _model_context_window(latest_model)

    # Wall-clock duration: first user prompt → last event seen.
    start = _parse_iso(first_user_ts)
    end = _parse_iso(last_event_ts)
    if start and end and end >= start:
        digest.duration_seconds = int((end - start).total_seconds())

    # Build the recap timeline. Each band is included only if its
    # source exists; "Last asked" is *not* a band — the user can read
    # the most recent turn directly in the transcript pane next door,
    # so the bar focuses on what's not visible there: the session's
    # earlier history.
    recap: list[RecapEntry] = []
    if first_user_text:
        recap.append(
            RecapEntry(
                label="Started",
                timestamp=first_user_ts,
                text=_clean_first_prompt(first_user_text),
            )
        )
    if compact_summary_text:
        recap.append(
            RecapEntry(
                label="Earlier",
                timestamp=compact_summary_ts,
                text=_truncate_lines(compact_summary_text.strip(), 8),
            )
        )
    if away_summary_text:
        recap.append(
            RecapEntry(
                label="Recently",
                timestamp=away_summary_ts,
                text=away_summary_text.strip(),
            )
        )
    digest.recap = recap

    return digest


__all__ = [
    "RecapEntry",
    "SessionDigest",
    "extract_digest",
]
